iceoryx_posh/source/popo/trigger_handle.cpp

```diff
@@ -21,27 +21,25 @@ namespace iox
 {
 namespace popo
 {
-/// explicitly implemented for MSVC and QNX
-TriggerHandle::TriggerHandle() noexcept
-{
-}
-
 TriggerHandle::TriggerHandle(ConditionVariableData& conditionVariableData,
-                             const cxx::function<void(uint64_t)> resetCallback,
+                             const cxx::function<void(uint64_t)>& resetCallback,
                              const uint64_t uniqueTriggerId) noexcept
     : m_conditionVariableDataPtr(&conditionVariableData)
     , m_resetCallback(resetCallback)
     , m_uniqueTriggerId(uniqueTriggerId)
 {
-    if (!m_resetCallback)
-    {
-        errorHandler(PoshError::POPO__TRIGGER_HANDLE_INVALID_RESET_CALLBACK, ErrorLevel::FATAL);
-    }
 }
 
 TriggerHandle::TriggerHandle(TriggerHandle&& rhs) noexcept
+    : m_conditionVariableDataPtr{[&] {
+        rhs.m_mutex.lock();
+        return rhs.m_conditionVariableDataPtr;
+    }()}
+    , m_resetCallback{std::move(rhs.m_resetCallback)}
+    , m_uniqueTriggerId{rhs.m_uniqueTriggerId}
 {
-    *this = std::move(rhs);
+    rhs.invalidate();
+    rhs.m_mutex.unlock();
```

Comment:
I cringed a bit when I saw the direct lock/unlock calls :laughing: 

Wasn't it possible to keep the logic as is?

Reply:
Sadly no since the `resetCallback` is now non nullable and has to be always set up. Making it an optional would somehow defeat the purpose in my opinion. So I have to set the first member, lock the mutex since this class must be threadsafe and then assign the remaining members.

After that I can invalidate the `rhs` and release the lock.

But with a little refactoring I would move all the internals into a struct `triggerHandleData_t` and pack this into a smart lock and the whole problem would vanish.

Reply:
We can also keep it as is. Up to you.

Reply:
Lets keep it that way for now.